### PR TITLE
fix: [CDS-50925]: add validation on workflow/pipeline name

### PR DESCRIPTION
**Pull request by rafaelmart1ns**
## Describe your changes

Add validation on the backend side to update the workflow name, pipeline name, and pipeline stage name containing a dot.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42885)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/resources/PipelineResource.java [hunk 1]
```diff
@@ -158,6 +166,19 @@ public RestResponse<Pipeline> update(
...
+      if (pipeline.getName().contains(CGConstants.DOT)) {
+        throw new InvalidArgumentsException("Pipeline name can only have a-z, A-Z, 0-9, -, (, ) and _", USER);
+      }
+      for (PipelineStage stage : pipeline.getPipelineStages()) {
+        if (stage.getName().contains(CGConstants.DOT)) {
+          throw new InvalidArgumentsException("Pipeline stage name can only have a-z, A-Z, 0-9, -, (, ) and _", USER);
+        }
+      }
+    }
+
```
Comment (lasdox) [inline 400-rest/src/main/java/software/wings/resources/PipelineResource.java:181]: lets move this to service layer and add unit test

Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/resources/PipelineResource.java]: it moved.

File: 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java [hunk 2]
```diff
@@ -242,7 +243,17 @@ public Pipeline update(Pipeline pipeline, boolean migration, boolean fromYaml) {
...
+            FeatureName.SPG_ENABLE_VALIDATION_WORKFLOW_PIPELINE_STAGE, pipeline.getAccountId())) {
+      if (pipeline.getName().contains(CGConstants.DOT)) {
+        throw new InvalidArgumentsException("Pipeline name can only have a-z, A-Z, 0-9, -, (, ) and _", USER);
+      }
+      for (PipelineStage stage : pipeline.getPipelineStages()) {
+        if (stage.getName().contains(CGConstants.DOT)) {
+          throw new InvalidArgumentsException("Pipeline stage name can only have a-z, A-Z, 0-9, -, (, ) and _", USER);
+        }
+      }
+    }
```
Comment (lasdox) [inline 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java:256]: lets just finally wrap this behind a method for better maintenance

**Comment by lasdox** [comment]
what about workflow validation?

File: 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java (hunk 2 above)
Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java]: done.

**Comment by rafaelmart1ns** [comment]
> what about workflow validation?

workflow has all validation, just the pipeline needed to validate the stage name.

**Review by lasdox — DISMISSED** [review]
> Reviewed 1 of 4 files at r1, 3 of 5 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42885)* status: 6 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by lasdox — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42885)* status: 6 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by rafaelmart1ns** [comment]
trigger ft1, ft2

**Review by vcalasansh — APPROVED** [review]
> Reviewed 1 of 4 files at r1, 2 of 5 files at r2, 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42885)* status: 2. unresolved discussions.

**Review by rafaelmart1ns — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 2 of 5 files at r2, 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42885)* status: 2. unresolved discussions.

**Review by lasdox — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42885)* status: :shipit: complete! ✓

**Pull request merged**